src/optimizer/transforms/char-case-insensitive-lowercase-transform.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2017-present Dmitry Soshnikov <dmitry.soshnikov@gmail.com>
+ */
+
+'use strict';
+
+/**
+ * Transforms case-insensitive regexp to lowercase
+ *
+ * /AaBbÏ/i -> /aabbï/i
+ */
+module.exports = {
+  _AZClassRanges: null,
+  _hasUFlag: false,
+  init(ast) {
+    this._AZClassRanges = new Set();
+    this._hasUFlag = ast.flags.includes('u');
+  },
+  shouldRun(ast) {
+    return ast.flags.includes('i');
+  },
+  Char(path) {
+    const {node, parent} = path;
+    if (isNaN(node.codePoint)) {
+      return;
+    }
+
+    // Engine support for case-insensitive matching without the u flag
+    // for characters above \u1000 does not seem reliable.
+    if (!this._hasUFlag && node.codePoint >= 0x1000) {
+      return;
+    }
+
+    if (parent.type === 'ClassRange') {
+      // The only class ranges we handle are inside A-Z
+      if (isAZClassRange(parent)) {
+        this._AZClassRanges.add(parent);
+      }
+      if (!this._AZClassRanges.has(parent)) {
```

Comment:
So this `if` goes after we do `this._AZClassRanges.add(parent);` on line 38. Does it mean the `return`  should be just in empty `else`?

```js
if (!isAZClassRange(parent)) {
  return;
}

this._AZClassRanges.add(parent);
```

Reply:
It's a bit more complicated, since the `isAZClassRange` test will return `false` when we process the `to` char of the range and the `from` char has already been lowercased.

Following your idea, I can still update as:

```js
// The only class ranges we handle must be inside A-Z.
// After the `from` char is processed, the isAZClassRange test
// will be false, so we use a Set to keep track
// of parents and process the `to` char.
if (!this._AZClassRanges.has(parent) && !isAZClassRange(parent)) {
  return;
}
this._AZClassRanges.add(parent);
```

Reply:
Gotcha! OK, sounds good.